store/tikv/coprocessor.go

```diff
@@ -943,6 +961,24 @@ func (worker *copIteratorWorker) handleCopResponse(bo *Backoffer, rpcCtx *RPCCon
 			}
 		}
 	}
+	if resp.pbResp.IsCacheHit && cacheValue != nil {
+		// Cache hit and is valid: use cached data as response data and we don't update the cache.
+		resp.pbResp.Data = cacheValue.Data
+	} else {
+		// Cache not hit or cache hit but not valid: update the cache if the response can be cached.
+		if cacheKey != nil && resp.pbResp.CacheLastVersion > 0 {
+			if worker.store.coprCache.CheckAdmission(resp.pbResp.Data.Size(), resp.detail.ProcessTime) {
+				newCacheValue := coprCacheValue{
+					// TODO: Verify that using the data buffer directly is still safe.
+					Data:              resp.pbResp.Data,
```

Comment:
Should use deep copy? I'm not sure if there is a problem with a shallow copy. @qw4990 PTAL

Reply:
I guess it is fine, because `pbResp.Data` is an encoded protobuf message bytes array. It may be only passed to the protobuf decoder later and I can't imagine that it would be manipulated to other kind of bytes.

Reply:
I think we should deep copy here. Executor will reuse the underlying memory of chunk.

Reply:
It is a protobuf encoded message. While decoding the protobuf message, will bytes fields be pointing at (part of) encoded bytes? @SunRunAway

Reply:
Yes. the underlying bytes will be referenced by `chunk.Chunk` used by executors without memory copy.

Reply:
`pbResp.Data` will be unmarshaled to `tipb.SelectResponse` [here](https://github.com/pingcap/tidb/blob/72cd3f2032256797dc7f341f94726e6f7ef2bbb2/distsql/select_result.go#L110).
Then `tipb.SelectResponse` will be decoded to `chunk.Chunk` [here](https://github.com/pingcap/tidb/blob/72cd3f2032256797dc7f341f94726e6f7ef2bbb2/distsql/select_result.go#L202)